### ISSUE TITLE
Jabber SASL can't login with password which contains non-ASCII chars

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5066]
|<img alt="binary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| binary reported|
|-|-|
|Trac ID|trac#5066|
|Type|defect|
|Created|2011-04-25 13:54:46Z|
|Branch|https://github.com/twisted/twisted/tree/xmpp-sasl-non-ascii-5066|

If I trying to connect to my jabber-account where password contains non-ASCII characters I get the following error: 

```
 File "/usr/lib64/python2.7/site-packages/twisted/words/protocols/jabber/sasl_mechanisms.py", line 218, in _gen_response
    a1 = "%s:%s:%s" % (H("%s:%s:%s" % (username, realm, password)),
exceptions.UnicodeDecodeError: 'ascii' codec can't decode byte 0xd0 in position 0: ordinal not in range(128)
```

It may seems to be strange because username and password are correctly encoded into "utf-8" with some lines before but realm is still in unicode and the whole string became to be unicode and exceptions is got.

To resolve this issue we must just encode realm too:

http://twistedmatrix.com/trac/browser/trunk/twisted/words/protocols/jabber/sasl_mechanisms.py#L207

after this line add: realm = realm.encode(charset)

Attachments:

* [5066_patch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/61b/61b129f8cf7d62fffb870024e320ac759133be10/66e1ff6f9104c4dc51a2bdd2a071f1e4a6ab9d4c.diff) (1986 bytes) - added by binary on 2012-08-17 14:31:58Z - 
* [5066_patch_2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/61b/61b129f8cf7d62fffb870024e320ac759133be10/b5f201c5cbce7262ceb642ab1dc176fcf310c1bd.diff) (5145 bytes) - added by binary on 2012-08-20 12:54:25Z - The second try
* [5066_patch_3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/61b/61b129f8cf7d62fffb870024e320ac759133be10/e1d9267cce3d9c0b1d86f2ae1cffe24f118241aa.diff) (5059 bytes) - added by binary on 2012-08-27 10:36:16Z - 
* [5066_patch_4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/61b/61b129f8cf7d62fffb870024e320ac759133be10/38add213a5edbceab48d12e86cea5d7f61a21025.diff) (5844 bytes) - added by malinoff on 2013-02-18 05:59:05Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5066 5066
type__defect defect
reporter__binary binary
priority__highest highest
milestone__ 
branch__branches_xmpp_sasl_non_ascii_5066 branches/xmpp-sasl-non-ascii-5066
branch_author__ralphm ralphm
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1303739686000000 1303739686000000
changetime__1389923323000000 1389923323000000
version__None None
owner__ralphm ralphm
cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to binary|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="binary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|binary removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set owner to @ralphm<br>@ralphm set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_6"></a>|@ralphm commented|
|-|-|

Thanks for looking into this! I am in conversation with some people in the XMPP Standards Foundation on this. The problem is that I'm not sure if `digest-uri` and `realm` may be directly encoded to the `encoding` or that the domain labels in them must be in Punycode. The various specs (like rfc:2831) are not entirely clear on this, unfortunately.

**Comment by twisted-trac** [comment]
|<img alt="binary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|binary commented|
|-|-|

I think that punycode would not be used anywhere except of DNS... But, sure thing, it would be better to check first.

Thanks for your work! Will wait about any progress.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_8"></a>|@ralphm set owner to binary<br>@ralphm set status to `new`|
|-|-|

So it turns out this is not a problem, most implementations seem to just encode it. My review notes:

 * Make it explicit in DigestMD5 that the `__init__` parameters are supposed to be `unicode` objects (by documenting them) and make the format strings there `unicode` strings, too.
 * About: `test_getResponseUnicode`.
   * It lacks a docstring. 
   * The new test doesn't actually test that those two parameters are in fact encoded. I.e. the only thing that is tested is that a unicode password now doesn't break anything. Could you add some tests around that?
 * Could you add a news file?

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="binary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|binary set owner to @ralphm|
|-|-|

Tried to follow your suggestions, please take a look.

I don't actually get how the NEWS file should look, but I can try if patch itself is good :)

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_10"></a>|@ralphm commented|
|-|-|

(In [35331]) Branching to 'xmpp-sasl-non-ascii-5066'

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_11"></a>|@ralphm set owner to binary|
|-|-|

I applied both patches to [branches/xmpp-sasl-non-ascii-5066.](branches/xmpp-sasl-non-ascii-5066.)

My comments:

 * I'd prefer the new test to be put at the bottom of the other ones testing `getResponse` (i.e. just before `test__parse`).
 * The change to a unicode string for the nested function `KD` is wrong as it calls `H` on it, and that in turn calls `md5`, which expects a bytes string.
 * The indent in the assignment for `reponse` was changed. This makes second argument to `KD` no longer line up below the first. I think it is easier to read that way, even though the spaces after the parentheses are not conforming to pep:8. It could be split up, but this is more or less literally how it is written in rfc:2831.
 * The duplication of the code from `_gen_response` in `test_getResponseUnicode` to compare the `response` parameter is undesirable. It doesn't sufficiently test `_gen_response` itself and I'd rather see that in a set of additional tests. E.g. it doesn't address the note on having to re-encode to ISO-8859-1 if all characters are in that set, even though 'UTF-8' was specified. There are more such horrible details, and this causes interoperability issues all over the place anyway. Don't waste your time with it, especially since Digest-MD5 has been obsoleted over a year ago (see rfc:6331). Or at least not as part of this ticket. For this test, you should just encode the results yourself and put the outcome in the test as literals, for both encodings.
 * See here for the documentation on writing [ReviewProcess#Newsfiles](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles)

**Comment by twisted-trac** [comment]
|<img alt="binary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|binary set owner to @ralphm|
|-|-|

Fixed all of these except of third because can't get it. I can't be sure that password was properly encoded anyway but calculate it myself, no?

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_13"></a>|@ralphm set owner to binary|
|-|-|

Replying to [binary](#note_12):
> Fixed all of these [..]

Thanks!

> [..] except of third because can't get it. I can't be sure that password was properly encoded anyway but calculate it myself, no?

The point I was trying to make is that the code in the test is a literal copy from the actual implementation that is being tested, including a comment. I suggested replacing it with a literal value, but you are right that you'd have to calculate it yourself. That may also not be the best thing to do.

It basically comes down to this: for every piece of code that we change in Twisted, we like it to have high quality tests. Doing that for the Digest-MD5 SASL mechanism would require quite some more tests, and even more changes to the actual implementation. Even if you would eventually come up with a reasonable solution, it might not actually interoperate with other implementations. The list of reasons in rfc:6331 are the sad realization of this.

So really I'd not encourage you to do all that, unless you really really need to. Might I suggest you work on adding a SCRAM implementation instead (in another ticket)? See [rfc:6120#security-mti-auth XMPP Core section 13.8.1] and [rfc:5802 SCRAM] for details. There's also this nice [whitepaper on SCRAM](http://www.isode.com/whitepapers/scram.html) by a prominent XMPP server implementation.

Coming back to this ticket, I think it would be sufficient to test that the directives `username`, `password` and `digest-uri` are properly encoded and leave `response` untested.

If you really want to test them anyway, convince yourself that the current implementation is in fact correct (for these values) and check against the literal values `0a810f0437e4eeada7e403bcea3f772c` and `4fd4b1d56845219cb4cb8dd13eaef214` respectively. Fortunately your choice of U+0418 CYRILLIC CAPITAL LETTER I completely dances around the ISO-8859-1 issue I mentioned.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_14"></a>|@ralphm commented|
|-|-|

I meant `realm' instead of `password`, of course.

**Comment by twisted-trac** [comment]
|<img alt="binary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|binary set owner to @ralphm|
|-|-|

Replying to [ralphm](#note_13):
> Replying to [binary](#note_12):
> > Fixed all of these [..]
> 
> Thanks!
> 

Thank you too! :)


> So really I'd not encourage you to do all that, unless you really really need to. Might I suggest you work on adding a SCRAM implementation instead (in another ticket)? See [rfc:6120#security-mti-auth XMPP Core section 13.8.1] and [rfc:5802 SCRAM] for details. There's also this nice [whitepaper on SCRAM](http://www.isode.com/whitepapers/scram.html) by a prominent XMPP server implementation.

I can try but not immediately because I am planning to go to a vacation soon, but that's an interesting task.

> 
> Coming back to this ticket, I think it would be sufficient to test that the directives `username`, `password` and `digest-uri` are properly encoded and leave `response` untested.
> 
> If you really want to test them anyway, convince yourself that the current implementation is in fact correct (for these values) and check against the literal values `0a810f0437e4eeada7e403bcea3f772c` and `4fd4b1d56845219cb4cb8dd13eaef214` respectively. Fortunately your choice of U+0418 CYRILLIC CAPITAL LETTER I completely dances around the ISO-8859-1 issue I mentioned.

But that's impossible to hardcode response values there because they are depend on random generated cnonce.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_16"></a>|@ralphm set owner to binary|
|-|-|

Replying to [binary](#note_15):
> Replying to [ralphm](#note_13):
> > Coming back to this ticket, I think it would be sufficient to test that the directives `username`, `realm` and `digest-uri` are properly encoded and leave `response` untested.
> > 
> > If you really want to test them anyway, convince yourself that the current implementation is in fact correct (for these values) and check against the literal values …
> 
> But that's impossible to hardcode response values there because they are depend on random generated cnonce.

Oh, of course. You have three options here:

 * Monkey patch `_gen_nonce`.
 * Factor out the core of this function in a new one that just calculates the response given `nonce`, `cnonce`, `nc` and the encoded `username`, `realm`, `password` and `digest_uri` parameters.

 This makes it possible to test the encoding and unparsing separately from the actual response generation. For testing `_gen_response` you'd then override the new function. I believe it to be more proper to test the outcome of this new function against pre-calculated values, if only as a sanity check. You could use the example from [rfc:2831#section-4 RFC 2831 section 4] (and mention this in the docstring of the new test).
 * Not testing `response` at all.

**Comment by twisted-trac** [comment]
|<img alt="binary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|binary commented|
|-|-|

Replying to [ralphm](#note_16):
> Replying to [binary](#note_15):
> > Replying to [ralphm](#note_13):
> > > Coming back to this ticket, I think it would be sufficient to test that the directives `username`, `realm` and `digest-uri` are properly encoded and leave `response` untested.
> > > 
> > > If you really want to test them anyway, convince yourself that the current implementation is in fact correct (for these values) and check against the literal values …
> > 
> > But that's impossible to hardcode response values there because they are depend on random generated cnonce.
> 
> Oh, of course. You have three options here:
> 
>  * Monkey patch `_gen_nonce`.
>  * Factor out the core of this function in a new one that just calculates the response given `nonce`, `cnonce`, `nc` and the encoded `username`, `realm`, `password` and `digest_uri` parameters.
> 
>  This makes it possible to test the encoding and unparsing separately from the actual response generation. For testing `_gen_response` you'd then override the new function. I believe it to be more proper to test the outcome of this new function against pre-calculated values, if only as a sanity check. You could use the example from [rfc:2831#section-4 RFC 2831 section 4] (and mention this in the docstring of the new test).
>  * Not testing `response` at all.

Can you provide me with an advice how to do that better? :) The problem is being to be unsolved for a too long time, I think but I don't really understand what is the problem with the current patch.

**Comment by twisted-trac** [comment]
|<img alt="binary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|binary set owner to @ralphm|
|-|-|

The next try.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_19"></a>|@ralphm commented|
|-|-|

I applied the new patch to the branch, and forced a build. Unfortunately it [fails](http://buildbot.twistedmatrix.com/builders/py-without-modules/builds/874). From the looks of it, this is because the arguments to `DigestMD5` in this test are already encoded, where its `__init__` assumes them to be unicode strings. I'll prepare a fix for it.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_20"></a>|@ralphm commented|
|-|-|

(In [37606]) Fix failing test, add tests for unicode domains and remove unused imports.

Re: [#5066](https://github.com/twisted/twisted/issues/5066).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_21"></a>|@ralphm removed owner|
|-|-|

Forced [another build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/xmpp-sasl-non-ascii-5066) with my changes. I also did some test by actually connecting to servers, with and without IDNs, and that seemed to be working as expected.

binary, could you check that my changes work nicely for your use, too?

Otherwise, I think this is ready for review, but with my changes I disqualified myself from doing that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun set owner to @ralphm|
|-|-|

Thanks!

  1. in the tests:
    1. in `test_getResponseNonAsciiRealm`, can you write the assertions about `directives` using a single `assertEqual` call?  That way if multiple items mismatch, a single test failure will report all of them.
    1. Can you expand the `test_getResponseNonAsciiRealm` docstring a little bit (ie, what aspect of the response is interesting in the case where the realm has non-ascii in it)?  Also, sadly, I think it is "An encoded" rather than "A encoded". ;)
    1. The docstring `test_getResponseNoRealmIDN` claims something "works", but doesn't explain what that means.  Can you expand it as well?  Also, I'm confused why it says something is happening "without realm".  "realm" seems to be the same as "host", and the test does supply a host to `DigestMD5`.
    1. `test_calculate_response` doesn't follow the naming convention (because `_calculate_response` doesn't follow it, I guess :/)
    1. and its docstring could be cleaned up a bit (in particular, it is usually a good idea to avoid having the active verb be "test").
    1. `test_calculate_response` also needs some more whitespace following it to separate it from `test__parse`.
  1. in the implementation:
    1. not introduced in this branch, but serv_type, serv_name, digest_uri are all named against the coding standard :/
    1. The `DigestMD5.__init__` parameters should have more than just `@type` I suppose
    1. Not mandatory, but it might be nice to use b"" for the definitely-these-are-bytes literals in the code being changed/added (instead of "")
    1. `_calculate_response` and `_gen_response` should document their return value/type at least, if not also all their parameters
    1. The new `realm = realm` line in `_gen_response` is probably not needed
  1. Can you add a link to the relevant specification (or if there is none, an archive of a mailing list post?) so it's clear what this code is trying to accomplish?

These basically seem to be cosmetic issues.  Please go ahead and merge after addressing them and forcing a build with green results.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [41290]) Address test-related review comments.

refs [#5066](https://github.com/twisted/twisted/issues/5066)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [41291]) Address implementation-related review comments.

refs [#5066](https://github.com/twisted/twisted/issues/5066)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [41293]) twistedchecker fixes

refs [#5066](https://github.com/twisted/twisted/issues/5066)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set status to `closed`|
|-|-|

(In [41294]) Merge xmpp-sasl-non-ascii-5066

Author: binary, ralphm, exarkun
Reviewer: ralphm, exarkun
Fixes: [#5066](https://github.com/twisted/twisted/issues/5066)

Add necessary encoding logic to the `DigestMD5` SASL mechanism in
`twisted.words.protocols.jabber.sasl_mechanisms` so that it supports
unicode arguments and encoded non-ASCII challenges.
